gtsam/hybrid/tests/testGaussianMixtureFactor.cpp

```diff
@@ -191,8 +197,422 @@ TEST(GaussianMixtureFactor, Error) {
+ * μ1 and μ2 for the Gaussian distribution around which we sample z.
+ *
+ * The resulting factor graph should eliminate to a Bayes net
+ * which represents a sigmoid function.
+ */
+TEST(GaussianMixtureFactor, GaussianMixtureModel) {
+  double mu0 = 1.0, mu1 = 3.0;
+  double sigma = 2.0;
+  auto model = noiseModel::Isotropic::Sigma(1, sigma);
+
+  DiscreteKey m(M(0), 2);
+  Key z = Z(0);
+
+  auto c0 = make_shared<GaussianConditional>(z, Vector1(mu0), I_1x1, model),
+       c1 = make_shared<GaussianConditional>(z, Vector1(mu1), I_1x1, model);
+
+  auto gm = new GaussianMixture({z}, {}, {m}, {c0, c1});
+  auto mixing = new DiscreteConditional(m, "0.5/0.5");
+
+  HybridBayesNet hbn;
+  hbn.emplace_back(gm);
+  hbn.emplace_back(mixing);
+
+  // The result should be a sigmoid.
+  // So should be m = 0.5 at z=3.0 - 1.0=2.0
+  VectorValues given;
+  given.insert(z, Vector1(mu1 - mu0));
+
+  HybridGaussianFactorGraph gfg = hbn.toFactorGraph(given);
+  HybridBayesNet::shared_ptr bn = gfg.eliminateSequential();
+
+  HybridBayesNet expected;
+  expected.emplace_back(new DiscreteConditional(m, "0.5/0.5"));
+
+  EXPECT(assert_equal(expected, *bn));
+}
+
+/* ************************************************************************* */
+/**
+ * Test a simple Gaussian Mixture Model represented as P(m)P(z|m)
+ * where m is a discrete variable and z is a continuous variable.
+ * m is binary and depending on m, we have 2 different means
+ * and covariances each for the
+ * Gaussian distribution around which we sample z.
+ *
+ * The resulting factor graph should eliminate to a Bayes net
+ * which represents a sigmoid function leaning towards
+ * the tighter covariance Gaussian.
+ */
+TEST(GaussianMixtureFactor, GaussianMixtureModel2) {
+  double mu0 = 1.0, mu1 = 3.0;
+  auto model0 = noiseModel::Isotropic::Sigma(1, 8.0);
+  auto model1 = noiseModel::Isotropic::Sigma(1, 4.0);
+
+  DiscreteKey m(M(0), 2);
+  Key z = Z(0);
+
+  auto c0 = make_shared<GaussianConditional>(z, Vector1(mu0), I_1x1, model0),
+       c1 = make_shared<GaussianConditional>(z, Vector1(mu1), I_1x1, model1);
+
+  auto gm = new GaussianMixture({z}, {}, {m}, {c0, c1});
+  auto mixing = new DiscreteConditional(m, "0.5/0.5");
+
+  HybridBayesNet hbn;
+  hbn.emplace_back(gm);
+  hbn.emplace_back(mixing);
+
+  // The result should be a sigmoid leaning towards model1
+  // since it has the tighter covariance.
+  // So should be m = 0.34/0.66 at z=3.0 - 1.0=2.0
+  VectorValues given;
+  given.insert(z, Vector1(mu1 - mu0));
+  HybridGaussianFactorGraph gfg = hbn.toFactorGraph(given);
+  HybridBayesNet::shared_ptr bn = gfg.eliminateSequential();
+
+  HybridBayesNet expected;
+  expected.emplace_back(
+      new DiscreteConditional(m, "0.338561851224/0.661438148776"));
+
+  EXPECT(assert_equal(expected, *bn));
+}
+
+/* ************************************************************************* */
+/**
+ * Test a model P(x0)P(z0|x0)p(x1|m1)p(z1|x1)p(m1).
```

Comment:
Puzzled. If this is a motion model, should it not depend on x0 ? i.e., p(x1|m1) -> p(x1|x0, m1) ?

Reply:
The example that we had on the board leads to this. For the motion model to be dependent on both x0 and x1, we need the measurement z being dependent on both x0 and x1 aka `P(z|x1, x0, m1)` which will give us `L(x1, x0, m1)` as the motion model.

Reply:
Still eagerly waiting for this example to be fixed :-)

Reply:
Should I kill this one in favor of `DifferentMeans`? If I change this one, they both will be the same example.

Reply:
I think we had decided on a HBN that has P(x1|x0,m), gets converted to FG, and there would be two versions: one with one measurement on x0, one with two measurements which would inform m.

Reply:
I just looked at differentMeans and I am confused:
I had expected
1. P(x0)P(x1|x1,m)P(m)P(z0|x0)
2. P(x0)P(x1|x1,m)P(m)P(z0|x0)P(z1|x1)
Just like we had on the whiteboard. P(z|x0,x1,m) is something different altogether which resembles (2). Also interesting and should probably give the same result if z==z1-z0, and that would be an interesting test. 

But indeed, if those *three* examples are there, HBN->HFG, then these ones are obsolete.